<!DOCTYPE html>

<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <meta name="mobile-web-app-capable" content="yes">
    <title>
        HackMD - Collaborative Markdown Knowledge Base
    </title>
    <link rel="icon" type="image/png" href="https://hackmd.io/favicon.png">
    <link rel="apple-touch-icon" href="https://hackmd.io/apple-touch-icon.png">

    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/ionicons/2.0.1/css/ionicons.min.css" integrity="sha256-3iu9jgsy9TpTwXKb7bNQzqWekRX7pPK+2OLj3R922fo=" crossorigin="anonymous" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/octicons/3.5.0/octicons.min.css" integrity="sha256-QiWfLIsCT02Sdwkogf6YMiQlj4NE84MKkzEMkZnMGdg=" crossorigin="anonymous" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/prism/1.5.1/themes/prism.min.css" integrity="sha256-vtR0hSWRc3Tb26iuN2oZHt3KRUomwTufNIf5/4oeCyg=" crossorigin="anonymous" />
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@hackmd/emojify.js@2.1.0/dist/css/basic/emojify.min.css" integrity="sha256-UOrvMOsSDSrW6szVLe8ZDZezBxh5IoIfgTwdNDgTjiU=" crossorigin="anonymous" />
    <style>
        @charset "UTF-8";@import url(https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,500,500i|Source+Code+Pro:300,400,500|Source+Sans+Pro:300,300i,400,400i,600,600i|Source+Serif+Pro&subset=latin-ext);.hljs{display:block;background:#fff;padding:.5em;color:#333;overflow-x:auto}.hljs-comment,.hljs-meta{color:#969896}.hljs-emphasis,.hljs-quote,.hljs-string,.hljs-strong,.hljs-template-variable,.hljs-variable{color:#df5000}.hljs-keyword,.hljs-selector-tag,.hljs-type{color:#a71d5d}.hljs-attribute,.hljs-bullet,.hljs-literal,.hljs-number,.hljs-symbol{color:#0086b3}.hljs-built_in,.hljs-builtin-name{color:#005cc5}.hljs-name,.hljs-section{color:#63a35c}.hljs-tag{color:#333}.hljs-attr,.hljs-selector-attr,.hljs-selector-class,.hljs-selector-id,.hljs-selector-pseudo,.hljs-title{color:#795da3}.hljs-addition{color:#55a532;background-color:#eaffea}.hljs-deletion{color:#bd2c00;background-color:#ffecec}.hljs-link{text-decoration:underline}.markdown-body{font-size:16px;line-height:1.5;word-wrap:break-word}.markdown-body:after,.markdown-body:before{display:table;content:""}.markdown-body:after{clear:both}.markdown-body>:first-child{margin-top:0!important}.markdown-body>:last-child{margin-bottom:0!important}.markdown-body a:not([href]){color:inherit;text-decoration:none}.markdown-body .absent{color:#c00}.markdown-body .anchor{float:left;padding-right:4px;margin-left:-20px;line-height:1}.markdown-body .anchor:focus{outline:none}.markdown-body blockquote,.markdown-body dl,.markdown-body ol,.markdown-body p,.markdown-body pre,.markdown-body table,.markdown-body ul{margin-top:0;margin-bottom:16px}.markdown-body hr{height:.25em;padding:0;margin:24px 0;background-color:#e7e7e7;border:0}.markdown-body blockquote{font-size:16px;padding:0 1em;color:#777;border-left:.25em solid #ddd}.markdown-body blockquote>:first-child{margin-top:0}.markdown-body blockquote>:last-child{margin-bottom:0}.markdown-body kbd,.popover kbd{display:inline-block;padding:3px 5px;font-size:11px;line-height:10px;color:#555;vertical-align:middle;background-color:#fcfcfc;border:1px solid #ccc;border-bottom-color:#bbb;border-radius:3px;box-shadow:inset 0 -1px 0 #bbb}.markdown-body .loweralpha{list-style-type:lower-alpha}.markdown-body h1,.markdown-body h2,.markdown-body h3,.markdown-body h4,.markdown-body h5,.markdown-body h6{margin-top:24px;margin-bottom:16px;font-weight:600;line-height:1.25}.markdown-body h1 .octicon-link,.markdown-body h2 .octicon-link,.markdown-body h3 .octicon-link,.markdown-body h4 .octicon-link,.markdown-body h5 .octicon-link,.markdown-body h6 .octicon-link{color:#000;vertical-align:middle;visibility:hidden}.markdown-body h1:hover .anchor,.markdown-body h2:hover .anchor,.markdown-body h3:hover .anchor,.markdown-body h4:hover .anchor,.markdown-body h5:hover .anchor,.markdown-body h6:hover .anchor{text-decoration:none}.markdown-body h1:hover .anchor .octicon-link,.markdown-body h2:hover .anchor .octicon-link,.markdown-body h3:hover .anchor .octicon-link,.markdown-body h4:hover .anchor .octicon-link,.markdown-body h5:hover .anchor .octicon-link,.markdown-body h6:hover .anchor .octicon-link{visibility:visible}.markdown-body h1 code,.markdown-body h1 tt,.markdown-body h2 code,.markdown-body h2 tt,.markdown-body h3 code,.markdown-body h3 tt,.markdown-body h4 code,.markdown-body h4 tt,.markdown-body h5 code,.markdown-body h5 tt,.markdown-body h6 code,.markdown-body h6 tt{font-size:inherit}.markdown-body h1{font-size:2em}.markdown-body h1,.markdown-body h2{padding-bottom:.3em;border-bottom:1px solid #eee}.markdown-body h2{font-size:1.5em}.markdown-body h3{font-size:1.25em}.markdown-body h4{font-size:1em}.markdown-body h5{font-size:.875em}.markdown-body h6{font-size:.85em;color:#777}.markdown-body ol,.markdown-body ul{padding-left:2em}.markdown-body ol.no-list,.markdown-body ul.no-list{padding:0;list-style-type:none}.markdown-body ol ol,.markdown-body ol ul,.markdown-body ul ol,.markdown-body ul ul{margin-top:0;margin-bottom:0}.markdown-body li>p{margin-top:16px}.markdown-body li+li{padding-top:.25em}.markdown-body dl{padding:0}.markdown-body dl dt{padding:0;margin-top:16px;font-size:1em;font-style:italic;font-weight:700}.markdown-body dl dd{padding:0 16px;margin-bottom:16px}.markdown-body table{display:block;width:100%;overflow:auto;word-break:normal;word-break:keep-all}.markdown-body table th{font-weight:700}.markdown-body table td,.markdown-body table th{padding:6px 13px;border:1px solid #ddd}.markdown-body table tr{background-color:#fff;border-top:1px solid #ccc}.markdown-body table tr:nth-child(2n){background-color:#f8f8f8}.markdown-body img{max-width:100%;box-sizing:content-box;background-color:#fff}.markdown-body img[align=right]{padding-left:20px}.markdown-body img[align=left]{padding-right:20px}.markdown-body .emoji{max-width:none;vertical-align:text-top;background-color:transparent}.markdown-body span.frame{display:block;overflow:hidden}.markdown-body span.frame>span{display:block;float:left;width:auto;padding:7px;margin:13px 0 0;overflow:hidden;border:1px solid #ddd}.markdown-body span.frame span img{display:block;float:left}.markdown-body span.frame span span{display:block;padding:5px 0 0;clear:both;color:#333}.markdown-body span.align-center{display:block;overflow:hidden;clear:both}.markdown-body span.align-center>span{display:block;margin:13px auto 0;overflow:hidden;text-align:center}.markdown-body span.align-center span img{margin:0 auto;text-align:center}.markdown-body span.align-right{display:block;overflow:hidden;clear:both}.markdown-body span.align-right>span{display:block;margin:13px 0 0;overflow:hidden;text-align:right}.markdown-body span.align-right span img{margin:0;text-align:right}.markdown-body span.float-left{display:block;float:left;margin-right:13px;overflow:hidden}.markdown-body span.float-left span{margin:13px 0 0}.markdown-body span.float-right{display:block;float:right;margin-left:13px;overflow:hidden}.markdown-body span.float-right>span{display:block;margin:13px auto 0;overflow:hidden;text-align:right}.markdown-body code,.markdown-body tt{padding:0;padding-top:.2em;padding-bottom:.2em;margin:0;font-size:85%;background-color:rgba(0,0,0,.04);border-radius:3px}.markdown-body code:after,.markdown-body code:before,.markdown-body tt:after,.markdown-body tt:before{letter-spacing:-.2em;content:"\00a0"}.markdown-body code br,.markdown-body tt br{display:none}.markdown-body del code{text-decoration:inherit}.markdown-body pre{word-wrap:normal}.markdown-body pre>code{padding:0;margin:0;font-size:100%;word-break:normal;white-space:pre;background:transparent;border:0}.markdown-body .highlight{margin-bottom:16px}.markdown-body .highlight pre{margin-bottom:0;word-break:normal}.markdown-body .highlight pre,.markdown-body pre{padding:16px;overflow:auto;font-size:85%;line-height:1.45;background-color:#f7f7f7;border-radius:3px}.markdown-body pre code,.markdown-body pre tt{display:inline;max-width:auto;padding:0;margin:0;overflow:visible;line-height:inherit;word-wrap:normal;background-color:transparent;border:0}.markdown-body pre code:after,.markdown-body pre code:before,.markdown-body pre tt:after,.markdown-body pre tt:before{content:normal}.markdown-body .csv-data td,.markdown-body .csv-data th{padding:5px;overflow:hidden;font-size:12px;line-height:1;text-align:left;white-space:nowrap}.markdown-body .csv-data .blob-line-num{padding:10px 8px 9px;text-align:right;background:#fff;border:0}.markdown-body .csv-data tr{border-top:0}.markdown-body .csv-data th{font-weight:700;background:#f8f8f8;border-top:0}.news .alert .markdown-body blockquote{padding:0 0 0 40px;border:0 none}.activity-tab .news .alert .commits,.activity-tab .news .markdown-body blockquote{padding-left:0}.task-list-item{list-style-type:none}.task-list-item label{font-weight:400}.task-list-item.enabled label{cursor:pointer}.task-list-item+.task-list-item{margin-top:3px}.task-list-item-checkbox{float:left;margin:.31em 0 .2em -1.3em!important;vertical-align:middle;cursor:default!important}.markdown-body{padding-top:40px;padding-bottom:40px;max-width:758px;overflow:visible!important;position:relative}.markdown-body .emoji{vertical-align:top}.markdown-body pre{border:inherit!important}.markdown-body code{color:inherit!important}.markdown-body pre code .wrapper{display:-moz-inline-flex;display:-ms-inline-flex;display:-o-inline-flex;display:inline-flex}.markdown-body pre code .gutter{float:left;overflow:hidden;-webkit-user-select:none;user-select:none}.markdown-body pre code .gutter.linenumber{text-align:right;position:relative;display:inline-block;cursor:default;z-index:4;padding:0 8px 0 0;min-width:20px;box-sizing:content-box;color:#afafaf!important;border-right:3px solid #6ce26c!important}.markdown-body pre code .gutter.linenumber>span:before{content:attr(data-linenumber)}.markdown-body pre code .code{float:left;margin:0 0 0 16px}.markdown-body .gist .line-numbers{border-left:none;border-top:none;border-bottom:none}.markdown-body .gist .line-data{border:none}.markdown-body .gist table{border-spacing:0;border-collapse:inherit!important}.markdown-body code[data-gist-id]{background:none;padding:0}.markdown-body code[data-gist-id]:after,.markdown-body code[data-gist-id]:before{content:""}.markdown-body code[data-gist-id] .blob-num{border:unset}.markdown-body code[data-gist-id] table{overflow:unset;margin-bottom:unset}.markdown-body code[data-gist-id] table tr{background:unset}.markdown-body[dir=rtl] pre{direction:ltr}.markdown-body[dir=rtl] code{direction:ltr;unicode-bidi:embed}.markdown-body .alert>p{margin-bottom:0}.markdown-body pre.abc,.markdown-body pre.flow-chart,.markdown-body pre.graphviz,.markdown-body pre.mermaid,.markdown-body pre.sequence-diagram,.markdown-body pre.vega{text-align:center;background-color:inherit;border-radius:0;white-space:inherit;overflow:visible}.markdown-body pre.abc>code,.markdown-body pre.flow-chart>code,.markdown-body pre.graphviz>code,.markdown-body pre.mermaid>code,.markdown-body pre.sequence-diagram>code,.markdown-body pre.vega>code{text-align:left}.markdown-body pre.abc>svg,.markdown-body pre.flow-chart>svg,.markdown-body pre.graphviz>svg,.markdown-body pre.mermaid>svg,.markdown-body pre.sequence-diagram>svg,.markdown-body pre.vega>svg{max-width:100%;height:100%}.markdown-body pre>code.wrap{white-space:pre-wrap;white-space:-moz-pre-wrap;white-space:-pre-wrap;white-space:-o-pre-wrap;word-wrap:break-word}.markdown-body .alert>p,.markdown-body .alert>ul{margin-bottom:0}.markdown-body summary{display:list-item}.markdown-body summary:focus{outline:none}.markdown-body details summary{cursor:pointer}.markdown-body details:not([open])>:not(summary){display:none}.markdown-body figure{margin:1em 40px}.markdown-body .mark,.markdown-body mark{background-color:#fff1a7}.vimeo,.youtube{cursor:pointer;display:table;text-align:center;background-position:50%;background-repeat:no-repeat;background-size:contain;background-color:#000;overflow:hidden}.vimeo,.youtube{position:relative;width:100%}.youtube{padding-bottom:56.25%}.vimeo img{width:100%;object-fit:contain;z-index:0}.youtube img{object-fit:cover;z-index:0}.vimeo iframe,.youtube iframe,.youtube img{width:100%;height:100%;position:absolute;top:0;left:0}.vimeo iframe,.youtube iframe{vertical-align:middle;z-index:1}.vimeo .icon,.youtube .icon{position:absolute;height:auto;width:auto;top:50%;left:50%;transform:translate(-50%,-50%);color:#fff;opacity:.3;transition:opacity .2s;z-index:0}.vimeo:hover .icon,.youtube:hover .icon{opacity:.6;transition:opacity .2s}.slideshare .inner,.speakerdeck .inner{position:relative;width:100%}.slideshare .inner iframe,.speakerdeck .inner iframe{position:absolute;top:0;bottom:0;left:0;right:0;width:100%;height:100%}.MJX_Assistive_MathML{display:none}#MathJax_Message{z-index:1000!important}.ui-infobar{position:relative;z-index:2;max-width:760px;margin:25px auto -25px;color:#777}.toc .invisable-node{list-style-type:none}.ui-toc{position:fixed;bottom:20px;z-index:998}.ui-toc.both-mode{margin-left:8px}.ui-toc.both-mode .ui-toc-label{height:40px;padding:10px 4px;border-top-left-radius:0;border-bottom-left-radius:0}.ui-toc-label{background-color:#e6e6e6;border:none;color:#868686;transition:opacity .2s}.ui-toc .open .ui-toc-label{opacity:1;color:#fff;transition:opacity .2s}.ui-toc-label:focus{opacity:.3;background-color:#ccc;color:#000}.ui-toc-label:hover{opacity:1;background-color:#ccc;transition:opacity .2s}.ui-toc-dropdown{margin-top:20px;margin-bottom:20px;padding-left:10px;padding-right:10px;max-width:45vw;width:25vw;max-height:70vh;overflow:auto;text-align:inherit}.ui-toc-dropdown>.toc{max-height:calc(70vh - 100px);overflow:auto}.ui-toc-dropdown[dir=rtl] .nav{padding-right:0;letter-spacing:.0029em}.ui-toc-dropdown a{overflow:hidden;text-overflow:ellipsis;white-space:pre}.ui-toc-dropdown .nav>li>a{display:block;padding:4px 20px;font-size:13px;font-weight:500;color:#767676}.ui-toc-dropdown .nav>li:first-child:last-child > ul,.ui-toc-dropdown .toc.expand ul{display:block}.ui-toc-dropdown .nav>li>a:focus,.ui-toc-dropdown .nav>li>a:hover{padding-left:19px;color:#000;text-decoration:none;background-color:transparent;border-left:1px solid #000}.ui-toc-dropdown[dir=rtl] .nav>li>a:focus,.ui-toc-dropdown[dir=rtl] .nav>li>a:hover{padding-right:19px;border-left:none;border-right:1px solid #000}.ui-toc-dropdown .nav>.active:focus>a,.ui-toc-dropdown .nav>.active:hover>a,.ui-toc-dropdown .nav>.active>a{padding-left:18px;font-weight:700;color:#000;background-color:transparent;border-left:2px solid #000}.ui-toc-dropdown[dir=rtl] .nav>.active:focus>a,.ui-toc-dropdown[dir=rtl] .nav>.active:hover>a,.ui-toc-dropdown[dir=rtl] .nav>.active>a{padding-right:18px;border-left:none;border-right:2px solid #000}.ui-toc-dropdown .nav .nav{display:none;padding-bottom:10px}.ui-toc-dropdown .nav>.active>ul{display:block}.ui-toc-dropdown .nav .nav>li>a{padding-top:1px;padding-bottom:1px;padding-left:30px;font-size:12px;font-weight:400}.ui-toc-dropdown[dir=rtl] .nav .nav>li>a{padding-right:30px}.ui-toc-dropdown .nav .nav>li>ul>li>a{padding-top:1px;padding-bottom:1px;padding-left:40px;font-size:12px;font-weight:400}.ui-toc-dropdown[dir=rtl] .nav .nav>li>ul>li>a{padding-right:40px}.ui-toc-dropdown .nav .nav>li>a:focus,.ui-toc-dropdown .nav .nav>li>a:hover{padding-left:29px}.ui-toc-dropdown[dir=rtl] .nav .nav>li>a:focus,.ui-toc-dropdown[dir=rtl] .nav .nav>li>a:hover{padding-right:29px}.ui-toc-dropdown .nav .nav>li>ul>li>a:focus,.ui-toc-dropdown .nav .nav>li>ul>li>a:hover{padding-left:39px}.ui-toc-dropdown[dir=rtl] .nav .nav>li>ul>li>a:focus,.ui-toc-dropdown[dir=rtl] .nav .nav>li>ul>li>a:hover{padding-right:39px}.ui-toc-dropdown .nav .nav>.active:focus>a,.ui-toc-dropdown .nav .nav>.active:hover>a,.ui-toc-dropdown .nav .nav>.active>a{padding-left:28px;font-weight:500}.ui-toc-dropdown[dir=rtl] .nav .nav>.active:focus>a,.ui-toc-dropdown[dir=rtl] .nav .nav>.active:hover>a,.ui-toc-dropdown[dir=rtl] .nav .nav>.active>a{padding-right:28px}.ui-toc-dropdown .nav .nav>.active>.nav>.active:focus>a,.ui-toc-dropdown .nav .nav>.active>.nav>.active:hover>a,.ui-toc-dropdown .nav .nav>.active>.nav>.active>a{padding-left:38px;font-weight:500}.ui-toc-dropdown[dir=rtl] .nav .nav>.active>.nav>.active:focus>a,.ui-toc-dropdown[dir=rtl] .nav .nav>.active>.nav>.active:hover>a,.ui-toc-dropdown[dir=rtl] .nav .nav>.active>.nav>.active>a{padding-right:38px}.markdown-body{font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Helvetica Neue,Helvetica,Roboto,Arial,sans-serif}html[lang^=ja] .markdown-body{font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Helvetica Neue,Helvetica,Roboto,Arial,Hiragino Kaku Gothic Pro,ヒラギノ角ゴ Pro W3,Osaka,Meiryo,メイリオ,MS Gothic,ＭＳ\ ゴシック,sans-serif}html[lang=zh-tw] .markdown-body{font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Helvetica Neue,Helvetica,Roboto,Arial,PingFang TC,Microsoft JhengHei,微軟正黑,sans-serif}html[lang=zh-cn] .markdown-body{font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Helvetica Neue,Helvetica,Roboto,Arial,PingFang SC,Microsoft YaHei,微软雅黑,sans-serif}html .markdown-body[lang^=ja]{font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Helvetica Neue,Helvetica,Roboto,Arial,Hiragino Kaku Gothic Pro,ヒラギノ角ゴ Pro W3,Osaka,Meiryo,メイリオ,MS Gothic,ＭＳ\ ゴシック,sans-serif}html .markdown-body[lang=zh-tw]{font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Helvetica Neue,Helvetica,Roboto,Arial,PingFang TC,Microsoft JhengHei,微軟正黑,sans-serif}html .markdown-body[lang=zh-cn]{font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Helvetica Neue,Helvetica,Roboto,Arial,PingFang SC,Microsoft YaHei,微软雅黑,sans-serif}html[lang^=ja] .ui-toc-dropdown{font-family:Source Sans Pro,Helvetica,Arial,Meiryo UI,MS PGothic,ＭＳ\ Ｐゴシック,sans-serif}html[lang=zh-tw] .ui-toc-dropdown{font-family:Source Sans Pro,Helvetica,Arial,Microsoft JhengHei UI,微軟正黑UI,sans-serif}html[lang=zh-cn] .ui-toc-dropdown{font-family:Source Sans Pro,Helvetica,Arial,Microsoft YaHei UI,微软雅黑UI,sans-serif}html .ui-toc-dropdown[lang^=ja]{font-family:Source Sans Pro,Helvetica,Arial,Meiryo UI,MS PGothic,ＭＳ\ Ｐゴシック,sans-serif}html .ui-toc-dropdown[lang=zh-tw]{font-family:Source Sans Pro,Helvetica,Arial,Microsoft JhengHei UI,微軟正黑UI,sans-serif}html .ui-toc-dropdown[lang=zh-cn]{font-family:Source Sans Pro,Helvetica,Arial,Microsoft YaHei UI,微软雅黑UI,sans-serif}.ui-affix-toc{position:fixed;top:0;max-width:15vw;max-height:70vh;overflow:auto}.back-to-top,.expand-toggle,.go-to-bottom{display:block;padding:4px 10px;margin-top:10px;margin-left:10px;font-size:12px;font-weight:500;color:#999}.back-to-top:focus,.back-to-top:hover,.expand-toggle:focus,.expand-toggle:hover,.go-to-bottom:focus,.go-to-bottom:hover{color:#563d7c;text-decoration:none}.back-to-top,.go-to-bottom{margin-top:0}.ui-user-icon{width:20px;height:20px;display:block;border-radius:3px;margin-top:2px;margin-bottom:2px;margin-right:5px;background-position:50%;background-repeat:no-repeat;background-size:cover}.ui-user-icon.small{width:18px;height:18px;display:inline-block;vertical-align:middle;margin:0 0 .2em}.ui-infobar>small>span{line-height:22px}.ui-infobar>small .dropdown{display:inline-block}.ui-infobar>small .dropdown a:focus,.ui-infobar>small .dropdown a:hover{text-decoration:none}.ui-published-note{color:#337ab7}.ui-published-note .fa{font-size:20px;vertical-align:top}.unselectable{-webkit-user-select:none;-o-user-select:none;user-select:none}.selectable{-webkit-user-select:text;-o-user-select:text;user-select:text}@media print{blockquote,div,img,pre,table{page-break-inside:avoid!important}a[href]:after{font-size:12px!important}}.markdown-body.slides{position:relative;z-index:1;color:#222}.markdown-body.slides:before{content:"";display:block;position:absolute;top:0;left:0;right:0;bottom:0;z-index:-1;background-color:currentColor;box-shadow:0 0 0 50vw}.markdown-body.slides section[data-markdown]{position:relative;margin-bottom:1.5em;background-color:#fff;text-align:center}.markdown-body.slides section[data-markdown] code{text-align:left}.markdown-body.slides section[data-markdown]:before{content:"";display:block;padding-bottom:56.23%}.markdown-body.slides section[data-markdown]>div:first-child{position:absolute;top:50%;left:1em;right:1em;transform:translateY(-50%);max-height:100%;overflow:hidden}.markdown-body.slides section[data-markdown]>ul{display:inline-block}.markdown-body.slides>section>section+section:after{content:"";position:absolute;top:-1.5em;right:1em;height:1.5em;border:3px solid #777}.site-ui-font{font-family:Source Sans Pro,Helvetica,Arial,sans-serif}html[lang^=ja] .site-ui-font{font-family:Source Sans Pro,Helvetica,Arial,Hiragino Kaku Gothic Pro,ヒラギノ角ゴ Pro W3,Osaka,Meiryo,メイリオ,MS Gothic,ＭＳ\ ゴシック,sans-serif}html[lang=zh-tw] .site-ui-font{font-family:Source Sans Pro,Helvetica,Arial,PingFang TC,Microsoft JhengHei,微軟正黑,sans-serif}html[lang=zh-cn] .site-ui-font{font-family:Source Sans Pro,Helvetica,Arial,PingFang SC,Microsoft YaHei,微软雅黑,sans-serif}body{font-smoothing:subpixel-antialiased!important;-webkit-font-smoothing:subpixel-antialiased!important;-moz-osx-font-smoothing:auto!important;text-shadow:0 0 1em transparent,1px 1px 1.2px rgba(0,0,0,.004);-webkit-overflow-scrolling:touch;letter-spacing:.025em;font-family:Source Sans Pro,Helvetica,Arial,sans-serif}html[lang^=ja] body{font-family:Source Sans Pro,Helvetica,Arial,Hiragino Kaku Gothic Pro,ヒラギノ角ゴ Pro W3,Osaka,Meiryo,メイリオ,MS Gothic,ＭＳ\ ゴシック,sans-serif}html[lang=zh-tw] body{font-family:Source Sans Pro,Helvetica,Arial,PingFang TC,Microsoft JhengHei,微軟正黑,sans-serif}html[lang=zh-cn] body{font-family:Source Sans Pro,Helvetica,Arial,PingFang SC,Microsoft YaHei,微软雅黑,sans-serif}abbr[title]{border-bottom:none;text-decoration:underline;-webkit-text-decoration:underline dotted;text-decoration:underline dotted}abbr[data-original-title],abbr[title]{cursor:help}body.modal-open{overflow-y:auto;padding-right:0!important}
    </style>
    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
    	<script src="https://cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv.min.js" integrity="sha256-3Jy/GbSLrg0o9y5Z5n1uw0qxZECH7C6OQpVBgNFYa0g=" crossorigin="anonymous"></script>
    	<script src="https://cdnjs.cloudflare.com/ajax/libs/respond.js/1.4.2/respond.min.js" integrity="sha256-g6iAfvZp+nDQ2TdTR/VVKJf3bGro4ub5fvWSWVRi2NE=" crossorigin="anonymous"></script>
		<script src="https://cdnjs.cloudflare.com/ajax/libs/es5-shim/4.5.9/es5-shim.min.js" integrity="sha256-8E4Is26QH0bD52WoQpcB+R/tcWQtpzlCojrybUd7Mxo=" crossorigin="anonymous"></script>
    <![endif]-->
</head>

<body>
    <div id="doc" class="markdown-body container-fluid comment-enabled comment-inner" data-hard-breaks="true"><h2 id="FAIRbioinfo-for-bioinformaticians-2020" data-id="FAIRbioinfo-for-bioinformaticians-2020"><a class="anchor hidden-xs" href="#FAIRbioinfo-for-bioinformaticians-2020" title="FAIRbioinfo-for-bioinformaticians-2020"><span class="octicon octicon-link"></span></a><span>FAIRbioinfo for bioinformaticians 2020</span></h2><p><span>Authors: C. Hernandez T. Denecker J.Sellier C. Toffano-Nioche</span></p><h3 id="Introduction" data-id="Introduction"><a class="anchor hidden-xs" href="#Introduction" title="Introduction"><span class="octicon octicon-link"></span></a><span>Introduction</span></h3><p><span>For this practical exercise on Snakemake we will:</span></p><ul>
<li><span>access to snakemake by the way of a docker container</span></li>
<li><span>access to analysis tools by the way of a conda environment (details about conda will be seen after)</span></li>
<li><span>create a first snakefile with one rule</span></li>
<li><span>add a second rule to create the first workflow</span></li>
</ul><p><span>During these exercises, we will execute several cycles:  executing snakemake, observing the result and improving the code. Each cycle correspond to an objective. Each code version will be noted exoX.smk with X a progressive digit.</span></p><h3 id="Exercice-setup" data-id="Exercice-setup"><a class="anchor hidden-xs" href="#Exercice-setup" title="Exercice-setup"><span class="octicon octicon-link"></span></a><span>Exercice setup</span></h3><p><span>We will access to Snakemake by running a docker image containing the conda tool (among other):</span></p><p><strong><span>docker miniconda3 (git 2.20.1 + conda 4.8.2)</span></strong></p><pre><code>docker run -i -t -v ${PWD}:/data  continuumio/miniconda3
</code></pre><p><span>And, we will access to the analysis tools thanks to a conda environment, envfair.yml designed for this exercise</span></p><p><strong><span>envfair.yml</span></strong></p><pre><code>channels:
    - conda-forge
    - bioconda
    - default
dependencies:
    - python =3.7.6 # specify  python  version 
    - snakemake-minimal=5.10.0 # workflow  manager
    - graphviz=2.42.3 # for  visualisation
    - xorg -libxrender
    - xorg -libxpm
    - wget=1.20.1 # for  downloading  files
    - fastqc=0.11.9 # for  the  RNAseq  analysis
    - bowtie2=2.4.1
    - samtools=1.101
    - subread=2.0.1
</code></pre><p><em><span>NB: The specification of the python version is not required but can  help  with  downstream  conflicts.</span></em></p><p><strong><span>Conda environment</span></strong></p><pre><code>conda  env  create  -n envfair  -f envfair.yml
conda  activate  envfair
</code></pre><h3 id="Objective-1-the-rule-concept-1-input" data-id="Objective-1-the-rule-concept-1-input"><a class="anchor hidden-xs" href="#Objective-1-the-rule-concept-1-input" title="Objective-1-the-rule-concept-1-input"><span class="octicon octicon-link"></span></a><span>Objective 1: the rule concept, 1 input</span></h3><p><span>Create a snakemake file named </span><em><span>exo1.smk</span></em><span> including the first step of the RNAseq workflow (the reads quality checking thank to the fastqc tool) on one of the RNAseq files</span></p><hr><p><em><strong><span>HINT</span></strong></em></p><ul>
<li><span>input file: SRR3099585_chr18.fastq.gz in a local directory of yours</span></li>
<li><span>fastqc access:  by running docker miniconda3+ activate the conda envfair environment</span></li>
<li><span>fastqc command:fastqc inputFileName --outdir ResultDirectory</span></li>
<li><span>the 2 fastqc result files (.zip &amp; .html) are named based on the prefix of input file</span></li>
</ul><hr><p><em><strong><span>Solution</span></strong></em></p><pre><code>rule fastqc:
  output:
    "FastQC/SRR3099585_chr18_fastqc.zip", 
    "FastQC/SRR3099585_chr18_fastqc.html"
  input:
    "Data/SRR3099585_chr18.fastq.gz"
  shell: "fastqc --outdir FastQC/ {input}"
</code></pre><p><strong><span>Snakemake run</span></strong></p><pre><code>snakemake  --snakefile  exo1.smk
</code></pre><p><em><strong><span>Oserve the results</span></strong></em></p><p><span>Look at the newly created FastQC directory:  Snakemake create needed directories.</span></p><h3 id="Objective-2-the-rule-concept-2-inputs" data-id="Objective-2-the-rule-concept-2-inputs"><a class="anchor hidden-xs" href="#Objective-2-the-rule-concept-2-inputs" title="Objective-2-the-rule-concept-2-inputs"><span class="octicon octicon-link"></span></a><span>Objective 2: the rule concept, 2 inputs</span></h3><p><span>Create snakemake file </span><em><span>exo2.smk</span></em><span>. Improve previous code: add a second input RNAseq file to the rule.</span></p><hr><p><em><strong><span>HINT</span></strong></em></p><ul>
<li><span>input file: SRR3099586_chr18.fastq.gz in a local directory of yours</span></li>
</ul><hr><p><em><strong><span>Solution</span></strong></em></p><pre><code>rule fastqc:
  output:
    "FastQC/SRR3099585_chr18_fastqc.zip", 
    "FastQC/SRR3099585_chr18_fastqc.html",
    "FastQC/SRR3099586_chr18_fastqc.zip", 
    "FastQC/SRR3099586_chr18_fastqc.html"
  input:
    "Data/SRR3099585_chr18.fastq.gz",
    "Data/SRR3099586_chr18.fastq.gz"
  shell: "fastqc --outdir FastQC/ {input}"
</code></pre><p><strong><span>Snakemake run</span></strong></p><pre><code># -s is the short form of the --snakefile option
snakemake -s exo2.smk
</code></pre><p><em><strong><span>Oserve the results</span></strong></em></p><p><span>Why does Snakemake reply “Nothing to be done”?</span><br>
<span>Two solutions:</span></p><ul>
<li><span>delete the FastQC directory (rm -Rf FastQC) and rerun the snakemake command</span></li>
<li><span>use the Snakemake --forcerules(-R) option:</span></li>
</ul><pre><code>snakemake -s exo2.smk -R fastqc
</code></pre><h3 id="Objective-3-expand-function-n-inputs" data-id="Objective-3-expand-function-n-inputs"><a class="anchor hidden-xs" href="#Objective-3-expand-function-n-inputs" title="Objective-3-expand-function-n-inputs"><span class="octicon octicon-link"></span></a><span>Objective 3: expand function, n inputs</span></h3><p><span>Create snakemake file </span><em><span>exo3.smk</span></em><span>. Improve previous code: add all the RNAseq files. Boring with writing all input and output file names? Use the expand()function to manage all the input RNAseq files at once.</span></p><hr><p><em><strong><span>HINT</span></strong></em></p><ul>
<li><span>create a Python list at the begining of the snakefile and containing all the basename of the input files (don’t include the ”.fastq.gz” suffix). Python list: list_name = [“item1”, “item2”, …, “itemN”]</span></li>
<li><span>replace the filename lists of the input and output directives by the expand()function</span></li>
</ul><hr><p><em><strong><span>Solution</span></strong></em></p><pre><code>SAMPLES = ["SRR3099585_chr18","SRR3099586_chr18","SRR3099587_chr18"]

rule fastqc:
  output:
    expand("FastQC/{sample}_fastqc.zip", sample = SAMPLES),
    expand("FastQC/{sample}_fastqc.html", sample = SAMPLES)
  input:
    expand("Data/{sample}.fastq.gz", sample = SAMPLES)
  shell: "fastqc --outdir FastQC/ {input}"
</code></pre><p><strong><span>Snakemake run</span></strong></p><pre><code>rm -Rf  FastQC/
snakemake  -s exo3.smk
</code></pre><h3 id="Objective-4-2-rules-no-link" data-id="Objective-4-2-rules-no-link"><a class="anchor hidden-xs" href="#Objective-4-2-rules-no-link" title="Objective-4-2-rules-no-link"><span class="octicon octicon-link"></span></a><span>Objective 4: 2 rules, no link</span></h3><p><span>Create snakemake file </span><em><span>exo4.smk</span></em><span>. Improve previous code: add a second rule, this will start a workflow. The second rule concerns the creation of an index file for the genome sequence (needed for the mapping step).  As the mapping tool is bowtie2,the index creation tool is bowtie2-build.</span></p><hr><p><em><strong><span>HINT</span></strong></em></p><ul>
<li><span>genome file (input): Data/O.tauri_genome.fna</span></li>
<li><span>use a Python list and the expand()function to manage the 6 index file names that will be created by the command:  “</span><em><span>.1.bt2" … "</span></em><span>.4.bt2”,"</span><em><span>.rev.1.bt2","</span></em><span>.rev.2.bt2"</span></li>
<li><span>command: </span><code>bowtie2-build genomeSequenceAccess indexAccessPrefix</code></li>
</ul><hr><p><em><strong><span>Solution</span></strong></em></p><pre><code>SAMPLES = ["SRR3099585_chr18","SRR3099586_chr18","SRR3099587_chr18"]
BIDX = ["1","2","3","4","rev.1","rev.2"]

rule genome_bwt2_index:
  output:
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  input:
    "Data/O.tauri_genome.fna"
  shell: "bowtie2-build {input} Tmp/Otauri"

rule fastqc:
  output:
    expand("FastQC/{sample}_fastqc.zip", sample = SAMPLES),
    expand("FastQC/{sample}_fastqc.html", sample = SAMPLES)
  input:
    expand("Data/{sample}.fastq.gz", sample = SAMPLES)
  shell: "fastqc --outdir FastQC/ {input}"
</code></pre><p><strong><span>Snakemake run</span></strong></p><pre><code>rm -Rf  FastQC/
snakemake  -s exo4.smk
</code></pre><p><em><strong><span>Observe the results</span></strong></em></p><p><span>Does Snakemake do the job? Why wasn’t the fastqc command launched?</span></p><p><em><strong><span>rule links</span></strong></em></p><p><span>Snakemake run the first rule and stop when the target files are present.Also, there is no link between the 2 rules because they concern two independent parts of the analysis.The solution is to add a rule that aggregate this 2 parts of the workflow</span></p><h3 id="Objective-5-The-target-rule" data-id="Objective-5-The-target-rule"><a class="anchor hidden-xs" href="#Objective-5-The-target-rule" title="Objective-5-The-target-rule"><span class="octicon octicon-link"></span></a><span>Objective 5: The target rule</span></h3><p><span>Create snakemake file </span><em><span>exo5.smk</span></em><span>. Improve previous code: add a “first” rule (rule all, target, …)  with the expected results for the 2rules (fastqc and genome_bwt2_index in its input: directive.</span></p><p><em><strong><span>Solution</span></strong></em></p><pre><code>SAMPLES = ["SRR3099585_chr18","SRR3099586_chr18","SRR3099587_chr18"]
BIDX = ["1","2","3","4","rev.1","rev.2"]

rule all:
  input:
    expand("FastQC/{sample}_fastqc.html", sample=SAMPLES),
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)

rule genome_bwt2_index:
  output:
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  input:
    "Data/O.tauri_genome.fna"
  shell: "bowtie2-build {input} Tmp/Otauri"

rule fastqc:
  output:
    expand("FastQC/{sample}_fastqc.zip", sample = SAMPLES),
    expand("FastQC/{sample}_fastqc.html", sample = SAMPLES)
  input:
    expand("Data/{sample}.fastq.gz", sample = SAMPLES)
  shell: "fastqc --outdir FastQC/ {input}"
</code></pre><p><strong><span>Snakemake run</span></strong></p><pre><code>rm -Rf  FastQC/
snakemake  -s exo5.smk -R all  fastqc
</code></pre><p><em><strong><span>Observe the results</span></strong></em><br>
<span>Does Snakemake do the job?</span></p><p><em><strong><span>Fastqc:  job or jobs?</span></strong></em><br>
<span>Look at more precisely the fastqc job.  We have many input files but snakemake launched only one fastqc job:</span></p><p><img src="https://github.com/chernan/FAIR_bioinfo_docs/blob/master/03_workflow/images/FAIR_ex1_o5_smk.png" alt="image"></p><p><span>It is because thefastqcrule is defined with a list of files and not for one unique file and because thefastqc tool accepts both a unique file as wellas a list of files.</span></p><h3 id="Objective-6-Running-n-individual-jobs" data-id="Objective-6-Running-n-individual-jobs"><a class="anchor hidden-xs" href="#Objective-6-Running-n-individual-jobs" title="Objective-6-Running-n-individual-jobs"><span class="octicon octicon-link"></span></a><span>Objective 6: Running n individual jobs</span></h3><p><span>Create snakemake file </span><em><span>exo6.smk</span></em><span>. Improve previous code: Thank to the all rule, all expected files are designated.  So we don’t need to give the fastqc rule a list anymore and we can replace it to manage only one file and all files one by one.  We will gain in power in system shaving more than one core.</span></p><hr><p><em><strong><span>HINT</span></strong></em><br>
<span>Replace the expand()function with a wildcard for one filename in the fastqc rule.</span></p><hr><p><strong><span>Solution</span></strong></p><pre><code>SAMPLES = ["SRR3099585_chr18","SRR3099586_chr18","SRR3099587_chr18"]
BIDX = ["1","2","3","4","rev.1","rev.2"]

rule all:
  input:
    expand("FastQC/{sample}_fastqc.html", sample=SAMPLES),
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)

rule genome_bwt2_index:
  output:
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  input:
    "Data/O.tauri_genome.fna"
  shell: "bowtie2-build {input} Tmp/Otauri"

rule fastqc:
  output:
    "FastQC/{sample}_fastqc.zip",
    "FastQC/{sample}_fastqc.html"
  input:
    "Data/{sample}.fastq.gz"
  shell: "fastqc --outdir FastQC/ {input}"
</code></pre><p><strong><span>Snakemake run</span></strong></p><pre><code>rm -Rf  FastQC/
snakemake  -s exo6.smk -R all fastqc
</code></pre><p><em><strong><span>Observe the results</span></strong></em><br>
<span>Now Snakemake did many fastqc jobs:</span></p><p><img src="https://github.com/chernan/FAIR_bioinfo_docs/blob/master/03_workflow/images/FAIR_ex1_o6_smk.png" alt="image"></p><p><span>But what happens to the runtime displays on the screen? To correct this, we will move the displays to a log file specific for each rule and each input file.</span></p><h3 id="Objective-7-Adding-log-file" data-id="Objective-7-Adding-log-file"><a class="anchor hidden-xs" href="#Objective-7-Adding-log-file" title="Objective-7-Adding-log-file"><span class="octicon octicon-link"></span></a><span>Objective 7: Adding log file</span></h3><p><span>Create snakemake file </span><em><span>exo7.smk</span></em><span>. Improve previous code:</span><br>
<span>In Unix systems, the output of a command is usually sent to two separate streams:  the normal output:  to Standard Out (stdout also “&gt;” in shell),and error messages:  to Standard Error (stderr, or “2&gt;” in shell).  To integrate stderr into the same log file as the stdout can be use “&amp;&gt;” instead of “&gt;”:</span><br>
<span>shell:  … &amp;&gt; {log}, but use with care when output files are printed tostdout (as often in shell comands). Redirect the stdout and stderr streams of the fastqc and bowtie2-build commands.</span></p><hr><p><em><strong><span>HINT</span></strong></em><br>
<span>For the bowtie2-build and fastqc rules, add the log: directive with two variables (log1 and log2) to redirect each streams.</span></p><hr><p><strong><span>Solution</span></strong></p><pre><code>SAMPLES = ["SRR3099585_chr18","SRR3099586_chr18","SRR3099587_chr18"]
BIDX = ["1","2","3","4","rev.1","rev.2"]

rule all:
  input:
    expand("FastQC/{sample}_fastqc.html", sample=SAMPLES),
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)

rule genome_bwt2_index:
  output:
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  input:
    "Data/O.tauri_genome.fna"
  log:
    log1="Logs/genome_bwt2_index.log1",
    log2="Logs/genome_bwt2_index.log2"
  shell: "bowtie2-build {input} Tmp/Otauri 1&gt;{log.log1} 2&gt;{log.log2}"

rule fastqc:
  output:
    "FastQC/{sample}_fastqc.zip",
    "FastQC/{sample}_fastqc.html"
  input:
    "Data/{sample}.fastq.gz"
  log:
    log1="Logs/{sample}_fastqc.log1",
    log2="Logs/{sample}_fastqc.log2"
  shell: "fastqc --outdir FastQC/ {input} 1&gt;{log.log1} 2&gt;{log.log2}"
</code></pre><p><strong><span>Snakemake run</span></strong></p><pre><code>rm -Rf  FastQC/ Tmp/ Logs/
snakemake  -s exo7.smk
</code></pre><h3 id="Setup-second-part" data-id="Setup-second-part"><a class="anchor hidden-xs" href="#Setup-second-part" title="Setup-second-part"><span class="octicon octicon-link"></span></a><span>Setup second part</span></h3><h4 id="Data" data-id="Data"><a class="anchor hidden-xs" href="#Data" title="Data"><span class="octicon octicon-link"></span></a><span>Data</span></h4><p><span>-g genome sequence access (including extention .fna, .fasta)</span><br>
<span>-a genome annotation access (inluding extention .gff)</span><br>
<span>-d RNAseq sample prefixnext args:  RNAseq sample prefix, no .fastq.gz extention</span></p><h4 id="Bash-command-line" data-id="Bash-command-line"><a class="anchor hidden-xs" href="#Bash-command-line" title="Bash-command-line"><span class="octicon octicon-link"></span></a><span>Bash command line</span></h4><pre><code>FAIR_initial_script.sh -g ../O.tauri_genome.fna -a ../O.tauri_annotation.gff -d ../SRR3099585_chr18 S*86_chr18 S*87_chr18 S*97 _chr18 S*98_chr18 S*99_chr18

</code></pre><h4 id="Script-in-3-main-blocks" data-id="Script-in-3-main-blocks"><a class="anchor hidden-xs" href="#Script-in-3-main-blocks" title="Script-in-3-main-blocks"><span class="octicon octicon-link"></span></a><span>Script in 3 main blocks</span></h4><ol>
<li><span>while  getops  do …  done</span></li>
<li><span>for  sample  in $* ; do …  done</span></li>
<li><span>creation  of the  result  file , counts.txt , with  paste , awk ,and sed  bash  commands</span></li>
</ol><h4 id="getops-block" data-id="getops-block"><a class="anchor hidden-xs" href="#getops-block" title="getops-block"><span class="octicon octicon-link"></span></a><span>getops block</span></h4><pre><code>while  getopts g:a:d: flag do
        case  $flag  in
            g)   genome=$OPTARG4echo  genome  is  $genome  ;;
            a)   annots=$OPTARG6echo  annotation  is  $annots  ;;
            d)   rnadir=$OPTARG8echo  RNAseq  path is  $rnadir  ;;
            :)   echo "L’option  $OPTARG  requiert  un  argument"10exit 1 ;;
            \?)  echo "$OPTARG : option  invalide"12exit 1 ;;
            esac
        done
    shift $((  OPTIND  - 1 ))   # shift  past  the  last  flag or argument

echo  samples  are $*

nbs=0;
for  sample  in $* ; do
    nbs=$(expr ${nbs} + 1)
    echo  traitement  of  sample ${sample}
    # --------- quality  control  of  reads
    if [ ! -d FastQC  ]; then
        mkdir  FastQC
    fi
    fastqc  --outdir  FastQC ${rnadir}${sample }.fastq.gz &gt;FastQC/${sample }.log 2&gt;&amp;1
    #---------- reads  mapping
    if [ ! -d Bwt2_index  ]; then
        mkdir  Bwt2_index
        bowtie2 -build ${genome} Bwt2_index/tauri &gt; Bwt2_index/Bwt2_index.log 2&gt;&amp;1
    fi
    bowtie2  -x Bwt2_index/tauri -U ${rnadir}${sample }. fastq.gz -S ${sample }.sam &gt; ${sample}_bowtie2.log 2&gt;&amp;1
    #---------- selection  and  format  modification
    samtools  view -b ${sample }.sam -o ${sample }.bam
    samtools  sort ${sample }.bam -o ${sample}_sort.bam
    samtools  index ${sample}_sort.bam
    #---------- counting  of  mapped  reads by gene
    featureCounts  -t gene -g ID -a ${annots} -s 2 -o ${sample}_ftc.txt ${sample}_sort.bam &gt; ${sample}_ftc.log 2&gt;&amp;1
done
</code></pre><pre><code># Count table block
paste *_ftc.txt &gt; ftc_tmp.txt
awk -v nb=${nbs} -v col=7 ’BEGIN{FS="\t"}{ ctmp=$1; for(i=col;i&lt;=nb*col;i=i+col){count=sprintf ("%s\t%s",ctmp ,$i);ctmp=count}; print  count}’ ftc_tmp.txt | sed 1d &gt; counts.txt
</code></pre><h3 id="Objective-8-Adding-a-configuration-file" data-id="Objective-8-Adding-a-configuration-file"><a class="anchor hidden-xs" href="#Objective-8-Adding-a-configuration-file" title="Objective-8-Adding-a-configuration-file"><span class="octicon octicon-link"></span></a><span>Objective 8: Adding a configuration file</span></h3><p><span>Create snakemake file </span><em><span>exo8.smk</span></em><span>. Improve previous code: add a configuration file, named RNAseq.yml, containing both the genome sequence and the annotation files manes, and the access to the Data directory. In the snakefile, change the configured variables (ex.  replace Data/ and genome by their config[] values).  The Python strings concatenation is +</span><br>
<span>Then, run snakemake with the </span><code>--configfile option</code></p><p><strong><span>Configuration file: exo8.yml</span></strong></p><pre><code>genome:
  O.tauri_genome.fna
annots:
  O.tauri_annotation.gff
dataDir:
  Data/
</code></pre><p><strong><span>Solution</span></strong></p><pre><code>SAMPLES = ["SRR3099585_chr18","SRR3099586_chr18","SRR3099587_chr18"]
BIDX = ["1","2","3","4","rev.1","rev.2"]

rule all:
  input:
    expand("FastQC/{sample}_fastqc.html", sample=SAMPLES),
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)

rule genome_bwt2_index:
  output:
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  input:
    config["dataDir"]+config["genome"]
  log:
    log1="Logs/genome_bwt2_index.log1",
    log2="Logs/genome_bwt2_index.log2"
  shell: "bowtie2-build {input} Tmp/Otauri 1&gt;{log.log1} 2&gt;{log.log2}"

rule fastqc:
  output: 
    "FastQC/{sample}_fastqc.zip",
    "FastQC/{sample}_fastqc.html"
  input: 
    config["dataDir"]+"{sample}.fastq.gz"
  log:
    log1="Logs/{sample}_fastqc.log1",
    log2="Logs/{sample}_fastqc.log2"
  shell: "fastqc --outdir FastQC/ {input} 1&gt;{log.log1} 2&gt;{log.log2}"
</code></pre><p><strong><span>Snakemake run</span></strong></p><pre><code>rm -Rf  FastQC/ Result/ Tmp/ Logs/ ; snakemake  -s exo8.smk--configfile  exo8.yml
</code></pre><h3 id="Objective-9-for-block" data-id="Objective-9-for-block"><a class="anchor hidden-xs" href="#Objective-9-for-block" title="Objective-9-for-block"><span class="octicon octicon-link"></span></a><span>Objective 9: for block</span></h3><p><span>Shell script</span></p><pre><code>nbs =0;
for sample in $* ; do
...
done
</code></pre><p><span>To manage all *.fastq.gz files in a directory, use the glob_wildcards() function. In ex2_o2.smk, replace the SAMPLES definition by:</span></p><pre><code>SAMPLES , = glob_wildcards ( config [" dataDir "]+"{ sample }. fastq .
gz ")
</code></pre><p><strong><span>Solution</span></strong></p><pre><code>SAMPLES, = glob_wildcards(config["dataDir"]+"{sample}.fastq.gz")
BIDX = ["1","2","3","4","rev.1","rev.2"]

rule all:
  input:
    expand("FastQC/{sample}_fastqc.html", sample=SAMPLES),
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)

rule genome_bwt2_index:
  output:
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  input:
    config["dataDir"]+config["genome"]
  log:
    log1="Logs/genome_bwt2_index.log1",
    log2="Logs/genome_bwt2_index.log2"
  shell: "bowtie2-build {input} Tmp/Otauri 1&gt;{log.log1} 2&gt;{log.log2}"

rule fastqc:
  output: 
    "FastQC/{sample}_fastqc.zip",
    "FastQC/{sample}_fastqc.html"
  input: 
    config["dataDir"]+"{sample}.fastq.gz"
  log:
    log1="Logs/{sample}_fastqc.log1",
    log2="Logs/{sample}_fastqc.log2"
  shell: "fastqc --outdir FastQC/ {input} 1&gt;{log.log1} 2&gt;{log.log2}"
</code></pre><h3 id="Objective-10-Quality-control-fastqc" data-id="Objective-10-Quality-control-fastqc"><a class="anchor hidden-xs" href="#Objective-10-Quality-control-fastqc" title="Objective-10-Quality-control-fastqc"><span class="octicon octicon-link"></span></a><span>Objective 10: Quality control, fastqc</span></h3><p><span>Shell script</span></p><pre><code>if [ ! -d FastQC ]; then
mkdir FastQC
fi
fastqc -- outdir FastQC ${ sample }. fastq .gz &gt; FastQC /${ sample
}. log 2 &gt;&amp;1
</code></pre><p><span>No more need to test the existence of a directory, it is created as needed.</span></p><p><strong><span>rule fastqc:</span></strong><br>
<span>This rule is already present in the snakefile</span></p><h3 id="Objective-11-Reads-mapping-bowtie2" data-id="Objective-11-Reads-mapping-bowtie2"><a class="anchor hidden-xs" href="#Objective-11-Reads-mapping-bowtie2" title="Objective-11-Reads-mapping-bowtie2"><span class="octicon octicon-link"></span></a><span>Objective 11: Reads mapping, bowtie2</span></h3><p><span>Shell script</span></p><pre><code>if [ ! -d Bwt2_index ]; then
mkdir Bwt2_index
bowtie2 - build ${ genome } Bwt2_index / tauri &gt; Bwt2_index /
Bwt2_index .log 2 &gt;&amp;1
fi
bowtie2 -x Bwt2_index / tauri -U ${ sample }. fastq .gz -S ${
sample }. sam &gt; ${ sample } _bowtie2 . log 2 &gt;&amp;1
</code></pre><p><strong><span>Solution</span></strong><br>
<span>add mapping rule + use {input[0]} + log err only</span></p><pre><code>SAMPLES, = glob_wildcards(config["dataDir"]+"{sample}.fastq.gz")
BIDX = ["1","2","3","4","rev.1","rev.2"]


rule bwt2_mapping:
  output:
    "Tmp/{sample}.sam"
  input:
    config["dataDir"]+"{sample}.fastq.gz",
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  log:
    "Logs/{sample}_bwt2_mapping.log"
  shell: "bowtie2 -x Tmp/Otauri -U {input[0]} -S {output} 2&gt; {log} "

rule genome_bwt2_index:
  output:
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  input:
    config["dataDir"]+config["genome"]
  log:
    log1="Logs/genome_bwt2_index.log1",
    log2="Logs/genome_bwt2_index.log2"
  shell: "bowtie2-build {input} Tmp/Otauri 1&gt;{log.log1} 2&gt;{log.log2}"

rule fastqc:
  output: 
    "FastQC/{sample}_fastqc.zip",
    "FastQC/{sample}_fastqc.html"
  input: 
    config["dataDir"]+"{sample}.fastq.gz"
  log:
    log1="Logs/{sample}_fastqc.log1",
    log2="Logs/{sample}_fastqc.log2"
  shell: "fastqc --outdir FastQC/ {input} 1&gt;{log.log1} 2&gt;{log.log2}"
</code></pre><p><span>Why some troubles?</span><br>
<span>The snakemake launch probably didn’t do what was expected. What have we forgotten?</span><br>
<span>We added a new rule to a snakemake but we didn’t manage the rule tree, their is no input-output link to include the new rule to the workflow.</span><br>
<span>We will do that by completing the input directive of the target rule (caution to respect the Python “list” structure, coma-separated).</span></p><p><strong><span>Solution</span></strong></p><pre><code>SAMPLES, = glob_wildcards(config["dataDir"]+"{sample}.fastq.gz")
BIDX = ["1","2","3","4","rev.1","rev.2"]

rule all:
  input:
    expand("FastQC/{sample}_fastqc.html", sample=SAMPLES),
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX),
    expand("Tmp/{sample}.sam", sample=SAMPLES)

rule bwt2_mapping:
  output:
    "Tmp/{sample}.sam"
  input:
    config["dataDir"]+"{sample}.fastq.gz",
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  log:
    "Logs/{sample}_bwt2_mapping.log"
  shell: "bowtie2 -x Tmp/Otauri -U {input[0]} -S {output} 2&gt; {log} "

rule genome_bwt2_index:
  output:
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  input:
    config["dataDir"]+config["genome"]
  log:
    log1="Logs/genome_bwt2_index.log1",
    log2="Logs/genome_bwt2_index.log2"
  shell: "bowtie2-build {input} Tmp/Otauri 1&gt;{log.log1} 2&gt;{log.log2}"

rule fastqc:
  output: 
    "FastQC/{sample}_fastqc.zip",
    "FastQC/{sample}_fastqc.html"
  input: 
    config["dataDir"]+"{sample}.fastq.gz"
  log:
    log1="Logs/{sample}_fastqc.log1",
    log2="Logs/{sample}_fastqc.log2"
  shell: "fastqc --outdir FastQC/ {input} 1&gt;{log.log1} 2&gt;{log.log2}"
</code></pre><h3 id="Objective-11-samtools" data-id="Objective-11-samtools"><a class="anchor hidden-xs" href="#Objective-11-samtools" title="Objective-11-samtools"><span class="octicon octicon-link"></span></a><span>Objective 11: samtools</span></h3><p><span>Shell script</span></p><pre><code>samtools sort -O bam -o ${ sample } _sort .bam ${ sample }. sam
samtools index ${ sample } _sort .bam
</code></pre><p><strong><span>Solution</span></strong></p><pre><code>SAMPLES, = glob_wildcards(config["dataDir"]+"{sample}.fastq.gz")
BIDX = ["1","2","3","4","rev.1","rev.2"]

rule all:
  input:
    expand("FastQC/{sample}_fastqc.html", sample=SAMPLES),
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX),
    expand("Tmp/{sample}.sam", sample=SAMPLES),
    expand("Result/{sample}_sort.bam.bai", sample=SAMPLES)

rule sam2bam_sort:
  output:
    bam="Result/{sample}_sort.bam",
    bai="Result/{sample}_sort.bam.bai"
  input:
    "Tmp/{sample}.sam"
  log:
    sort="Logs/{sample}_sam2bam_sort.log",
    index="Logs/{sample}_bam2bai.log"
  shell: 
    "samtools sort -O bam -o {output.bam} {input} 2&gt; {log.sort} ;"
    "samtools index {output.bam} 2&gt; {log.index}"

rule bwt2_mapping:
  output:
    "Tmp/{sample}.sam"
  input:
    config["dataDir"]+"{sample}.fastq.gz",
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  log:
    "Logs/{sample}_bwt2_mapping.log"
  shell: "bowtie2 -x Tmp/Otauri -U {input[0]} -S {output} 2&gt; {log} "

rule genome_bwt2_index:
  output:
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  input:
    config["dataDir"]+config["genome"]
  log:
    log1="Logs/genome_bwt2_index.log1",
    log2="Logs/genome_bwt2_index.log2"
  shell: "bowtie2-build {input} Tmp/Otauri 1&gt;{log.log1} 2&gt;{log.log2}"

rule fastqc:
  output: 
    "FastQC/{sample}_fastqc.zip",
    "FastQC/{sample}_fastqc.html"
  input: 
    config["dataDir"]+"{sample}.fastq.gz"
  log:
    log1="Logs/{sample}_fastqc.log1",
    log2="Logs/{sample}_fastqc.log2"
  shell: "fastqc --outdir FastQC/ {input} 1&gt;{log.log1} 2&gt;{log.log2}"
</code></pre><h3 id="Objective-12-FeatureCount" data-id="Objective-12-FeatureCount"><a class="anchor hidden-xs" href="#Objective-12-FeatureCount" title="Objective-12-FeatureCount"><span class="octicon octicon-link"></span></a><span>Objective 12: FeatureCount</span></h3><p><span>Shell script</span></p><pre><code>featureCounts -t gene -g ID -a ${ annots } -s  -o ${ sample }
_ftc . txt ${ sample } _sort . bam &gt; ${ sample } _ftc . log 2 &gt;&amp;1
</code></pre><p><strong><span>Solution</span></strong><br>
<span>add featureCount rule, params directive</span></p><pre><code>SAMPLES, = glob_wildcards(config["dataDir"]+"{sample}.fastq.gz")
BIDX = ["1","2","3","4","rev.1","rev.2"]

rule all:
  input:
    expand("FastQC/{sample}_fastqc.html", sample=SAMPLES),
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX),
    expand("Tmp/{sample}.sam", sample=SAMPLES),
    expand("Result/{sample}_sort.bam.bai", sample=SAMPLES),
    expand("Tmp/{sample}_ftc.txt", sample=SAMPLES)

rule counting:
  output:
    "Tmp/{sample}_ftc.txt"
  input:
    bam="Result/{sample}_sort.bam",
    annot=config["dataDir"]+config["annots"]
  params: t="gene", g="ID", s="2"
    "Logs/{sample}_counts.log"
  shell:
    "featureCounts -t {params.t} -g {params.g} -a {input.annot} -s {params.s} -o {output} {input.bam} &amp;&gt; {log}"

rule sam2bam_sort:
  output:
    bam="Result/{sample}_sort.bam",
    bai="Result/{sample}_sort.bam.bai"
  input:
    "Tmp/{sample}.sam"
  log:
    sort="Logs/{sample}_sam2bam_sort.log",
    index="Logs/{sample}_bam2bai.log"
  shell: 
    "samtools sort -O bam -o {output.bam} {input} 2&gt; {log.sort} ;"
    "samtools index {output.bam} 2&gt; {log.index}"


rule bwt2_mapping:
  output:
    "Tmp/{sample}.sam"
  input:
    config["dataDir"]+"{sample}.fastq.gz",
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  log:
    "Logs/{sample}_bwt2_mapping.log"
  shell: "bowtie2 -x Tmp/Otauri -U {input[0]} -S {output} 2&gt; {log} "

rule genome_bwt2_index:
  output:
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  input:
    config["dataDir"]+config["genome"]
  log:
    log1="Logs/genome_bwt2_index.log1",
    log2="Logs/genome_bwt2_index.log2"
  shell: "bowtie2-build {input} Tmp/Otauri 1&gt;{log.log1} 2&gt;{log.log2}"

rule fastqc:
  output: 
    "FastQC/{sample}_fastqc.zip",
    "FastQC/{sample}_fastqc.html"
  input: 
    config["dataDir"]+"{sample}.fastq.gz"
  log:
    log1="Logs/{sample}_fastqc.log1",
    log2="Logs/{sample}_fastqc.log2"
  shell: "fastqc --outdir FastQC/ {input} 1&gt;{log.log1} 2&gt;{log.log2}"
</code></pre><h3 id="Objective-13-Count-matrix-creation" data-id="Objective-13-Count-matrix-creation"><a class="anchor hidden-xs" href="#Objective-13-Count-matrix-creation" title="Objective-13-Count-matrix-creation"><span class="octicon octicon-link"></span></a><span>Objective 13: Count matrix creation</span></h3><p><span>Shell script</span></p><pre><code>paste * _ftc .txt &gt; counts_tmp .txt
awk -v nb=${ nb_sample } 'BEGIN {FS ="\ t"}{ count_tmp =$1; for (i
=7;i &lt;= nb *7; i=i+7) { count = sprintf ("% s\t%s", count_tmp ,$i);
count_tmp = count }; print count }' counts_tmp .txt | sed 1d &gt;
counts . txt
</code></pre><hr><p><em><strong><span>HINT</span></strong></em><br>
<span>Create 2 rules to manage some files aggregation to one result file:</span></p><ul>
<li><span>rule extract_counts: extract geneID and counts in individual files</span></li>
<li><span>rule matrix_counts: paste these files</span></li>
</ul><hr><p><strong><span>Solution</span></strong><br>
<span>add matrix count rule (2 rules to manage many files aggregation to one)</span></p><pre><code>SAMPLES, = glob_wildcards(config["dataDir"]+"{sample}.fastq.gz")
BIDX = ["1","2","3","4","rev.1","rev.2"]

rule all:
  input:
    expand("FastQC/{sample}_fastqc.html", sample=SAMPLES),
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX),
    expand("Tmp/{sample}.sam", sample=SAMPLES),
    expand("Result/{sample}_sort.bam.bai", sample=SAMPLES),
    expand("Tmp/{sample}_ftc.txt", sample=SAMPLES),
    "Result/counts_matrix.txt"

rule matrix_counts:
  output:
    "Result/counts_matrix.txt"
  input:
    countfile=expand("Tmp/{sample}_ftc7.txt", sample=SAMPLES),
    geneID=expand("Tmp/{sample}_ftc1.txt", sample=SAMPLES)
  log:
    "Logs/matrix_counts.log"
  shell:
    """cp {input.geneID[0]} Tmp/ftc_geneID.txt ; paste Tmp/ftc_geneID.txt {input.countfile} &gt; {output} &amp;&gt; {log}"""

rule extract_counts:
  output:
    col7="Tmp/{sample}_ftc7.txt",
    col1="Tmp/{sample}_ftc1.txt"
  input:
    "Tmp/{sample}_ftc.txt"
  log:
    "Logs/{sample}_extract_counts.log"
  shell: """cut -f 7- {input} | sed 1d &gt; {output.col7} &amp;&gt; {log} ; cut -f 1 {input} | sed 1d &gt; {output.col1} """

rule counting:
  output:
    "Tmp/{sample}_ftc.txt"
  input:
    bam="Result/{sample}_sort.bam",
    annot=config["dataDir"]+config["annots"]
  params: t="gene", g="ID", s="2"
  log:
    "Logs/{sample}_counts.log"
  shell:
    "featureCounts -t {params.t} -g {params.g} -a {input.annot} -s {params.s} -o {output} {input.bam} &amp;&gt; {log}"

rule sam2bam_sort:
  output:
    bam="Result/{sample}_sort.bam",
    bai="Result/{sample}_sort.bam.bai"
  input:
    "Tmp/{sample}.sam"
  log:
    sort="Logs/{sample}_sam2bam_sort.log",
    index="Logs/{sample}_bam2bai.log"
  shell: 
    "samtools sort -O bam -o {output.bam} {input} 2&gt; {log.sort} ;"
    "samtools index {output.bam} 2&gt; {log.index}"


rule bwt2_mapping:
  output:
    "Tmp/{sample}.sam"
  input:
    config["dataDir"]+"{sample}.fastq.gz",
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  log:
    "Logs/{sample}_bwt2_mapping.log"
  shell: "bowtie2 -x Tmp/Otauri -U {input[0]} -S {output} 2&gt; {log} "

rule genome_bwt2_index:
  output:
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  input:
    config["dataDir"]+config["genome"]
  log:
    log1="Logs/genome_bwt2_index.log1",
    log2="Logs/genome_bwt2_index.log2"
  shell: "bowtie2-build {input} Tmp/Otauri 1&gt;{log.log1} 2&gt;{log.log2}"

rule fastqc:
  output: 
    "FastQC/{sample}_fastqc.zip",
    "FastQC/{sample}_fastqc.html"
  input: 
    config["dataDir"]+"{sample}.fastq.gz"
  log:
    log1="Logs/{sample}_fastqc.log1",
    log2="Logs/{sample}_fastqc.log2"
  shell: "fastqc --outdir FastQC/ {input} 1&gt;{log.log1} 2&gt;{log.log2}"
</code></pre><p><span>ex2_o7.yml</span></p><pre><code>channels:
  - conda-forge
  - bioconda
  - main
  - r
  - default
dependencies:
  # Specify python version (not required but can help with downstream conflicts)
  - python=3.7.6
  # The workflow manager
  - snakemake-minimal=5.10.0
  # For visualizing workflows
  - graphviz=2.42.3
  - xorg-libxrender
  - xorg-libxpm
  # For downloading files
  - wget=1.20.1
  # For the RNAseq analysis
  - fastqc=0.11.9
  - bowtie2=2.4.1
  - samtools=1.10
  - subread=2.0.1
</code></pre><p><span>ex2_o7.smk</span><br>
<span>add conda: “xx.yml”</span></p><pre><code>SAMPLES, = glob_wildcards(config["dataDir"]+"{sample}.fastq.gz")
BIDX = ["1","2","3","4","rev.1","rev.2"]

rule all:
  input:
    expand("FastQC/{sample}_fastqc.html", sample=SAMPLES),
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX),
    expand("Tmp/{sample}.sam", sample=SAMPLES),
    expand("Result/{sample}_sort.bam.bai", sample=SAMPLES),
    expand("Tmp/{sample}_ftc.txt", sample=SAMPLES),
    "Result/counts_matrix.txt"

rule matrix_counts:
  output:
    "Result/counts_matrix.txt"
  input:
    countfile=expand("Tmp/{sample}_ftc7.txt", sample=SAMPLES),
    geneID=expand("Tmp/{sample}_ftc1.txt", sample=SAMPLES)
  log:
    "Logs/matrix_counts.log"
  shell:
    """cp {input.geneID[0]} Tmp/ftc_geneID.txt ; paste Tmp/ftc_geneID.txt {input.countfile} &gt; {output} &amp;&gt; {log}"""

rule extract_counts:
  output:
    col7="Tmp/{sample}_ftc7.txt",
    col1="Tmp/{sample}_ftc1.txt"
  input:
    "Tmp/{sample}_ftc.txt"
  log:
    "Logs/{sample}_extract_counts.log"
  shell: """cut -f 7- {input} | sed 1d &gt; {output.col7} &amp;&gt; {log} ; cut -f 1 {input} | sed 1d &gt; {output.col1} """

rule counting:
  output:
    "Tmp/{sample}_ftc.txt"
  input:
    bam="Result/{sample}_sort.bam",
    annot=config["dataDir"]+config["annots"]
  params: t="gene", g="ID", s="2"
  log:
    "Logs/{sample}_counts.log"
  conda: "condaEnv4SmkRules/counting.yml"
  shell:
    "featureCounts -t {params.t} -g {params.g} -a {input.annot} -s {params.s} -o {output} {input.bam} &amp;&gt; {log}"

rule sam2bam_sort:
  output:
    bam="Result/{sample}_sort.bam",
    bai="Result/{sample}_sort.bam.bai"
  input:
    "Tmp/{sample}.sam"
  log:
    sort="Logs/{sample}_sam2bam_sort.log",
    index="Logs/{sample}_bam2bai.log"
  conda: "condaEnv4SmkRules/sam2bam_sort.yml"
  shell: 
    "samtools sort -O bam -o {output.bam} {input} 2&gt; {log.sort} ;"
    "samtools index {output.bam} 2&gt; {log.index}"


rule bwt2_mapping:
  output:
    "Tmp/{sample}.sam"
  input:
    config["dataDir"]+"{sample}.fastq.gz",
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  log:
    "Logs/{sample}_bwt2_mapping.log"
  conda: "condaEnv4SmkRules/bwt2_mapping.yml"
  shell: "bowtie2 -x Tmp/Otauri -U {input[0]} -S {output} 2&gt; {log} "

rule genome_bwt2_index:
  output:
    expand("Tmp/Otauri.{ext}.bt2", ext=BIDX)
  input:
    config["dataDir"]+config["genome"]
  log:
    log1="Logs/genome_bwt2_index.log1",
    log2="Logs/genome_bwt2_index.log2"
  conda: "condaEnv4SmkRules/bwt2_mapping.yml"
  shell: "bowtie2-build {input} Tmp/Otauri 1&gt;{log.log1} 2&gt;{log.log2}"

rule fastqc:
  output: 
    "FastQC/{sample}_fastqc.zip",
    "FastQC/{sample}_fastqc.html"
  input: 
    config["dataDir"]+"{sample}.fastq.gz"
  log:
    log1="Logs/{sample}_fastqc.log1",
    log2="Logs/{sample}_fastqc.log2"
  conda: "condaEnv4SmkRules/fastqc.yml"
  shell: "fastqc --outdir FastQC/ {input} 1&gt;{log.log1} 2&gt;{log.log2}"
</code></pre><h3 id="DEseq2" data-id="DEseq2"><a class="anchor hidden-xs" href="#DEseq2" title="DEseq2"><span class="octicon octicon-link"></span></a><span>DEseq2</span></h3><p><span>Clean delete and re-run</span></p><pre><code>cp ex2_o8 . smk RNAseq_analysis . smk
cp ex2_o1 . yml RNAseq_analysis_smkEnv . yml
rm -Rf FastQC / Results / Logs / Tmp/
snakemake -s RNAseq_analysis .smk -- configfile
RNAseq_analysis_smkEnv .yml
</code></pre></div>
    <div class="ui-toc dropup unselectable hidden-print" style="display:none;">
        <div class="pull-right dropdown">
            <a id="tocLabel" class="ui-toc-label btn btn-default" data-toggle="dropdown" href="#" role="button" aria-haspopup="true" aria-expanded="false" title="Table of content">
                <i class="fa fa-bars"></i>
            </a>
            <ul id="ui-toc" class="ui-toc-dropdown dropdown-menu" aria-labelledby="tocLabel">
                <div class="toc"><ul class="nav">
<li class=""><a href="#FAIRbioinfo-for-bioinformaticians-2020" title="FAIRbioinfo for bioinformaticians 2020">FAIRbioinfo for bioinformaticians 2020</a><ul class="nav">
<li class=""><a href="#Introduction" title="Introduction">Introduction</a></li>
<li class=""><a href="#Exercice-setup" title="Exercice setup">Exercice setup</a></li>
<li><a href="#Objective-1-the-rule-concept-1-input" title="Objective 1: the rule concept, 1 input">Objective 1: the rule concept, 1 input</a></li>
<li><a href="#Objective-2-the-rule-concept-2-inputs" title="Objective 2: the rule concept, 2 inputs">Objective 2: the rule concept, 2 inputs</a></li>
<li><a href="#Objective-3-expand-function-n-inputs" title="Objective 3: expand function, n inputs">Objective 3: expand function, n inputs</a></li>
<li><a href="#Objective-4-2-rules-no-link" title="Objective 4: 2 rules, no link">Objective 4: 2 rules, no link</a></li>
<li class=""><a href="#Objective-5-The-target-rule" title="Objective 5: The target rule">Objective 5: The target rule</a></li>
<li><a href="#Objective-6-Running-n-individual-jobs" title="Objective 6: Running n individual jobs">Objective 6: Running n individual jobs</a></li>
<li><a href="#Objective-7-Adding-log-file" title="Objective 7: Adding log file">Objective 7: Adding log file</a></li>
<li><a href="#Setup-second-part" title="Setup second part">Setup second part</a><ul class="nav">
<li><a href="#Data" title="Data">Data</a></li>
<li><a href="#Bash-command-line" title="Bash command line">Bash command line</a></li>
<li><a href="#Script-in-3-main-blocks" title="Script in 3 main blocks">Script in 3 main blocks</a></li>
<li><a href="#getops-block" title="getops block">getops block</a></li>
</ul>
</li>
<li><a href="#Objective-8-Adding-a-configuration-file" title="Objective 8: Adding a configuration file">Objective 8: Adding a configuration file</a></li>
<li><a href="#Objective-9-for-block" title="Objective 9: for block">Objective 9: for block</a></li>
<li><a href="#Objective-10-Quality-control-fastqc" title="Objective 10: Quality control, fastqc">Objective 10: Quality control, fastqc</a></li>
<li><a href="#Objective-11-Reads-mapping-bowtie2" title="Objective 11: Reads mapping, bowtie2">Objective 11: Reads mapping, bowtie2</a></li>
<li><a href="#Objective-11-samtools" title="Objective 11: samtools">Objective 11: samtools</a></li>
<li><a href="#Objective-12-FeatureCount" title="Objective 12: FeatureCount">Objective 12: FeatureCount</a></li>
<li><a href="#Objective-13-Count-matrix-creation" title="Objective 13: Count matrix creation">Objective 13: Count matrix creation</a></li>
<li><a href="#DEseq2" title="DEseq2">DEseq2</a></li>
</ul>
</li>
</ul>
</div><div class="toc-menu"><a class="expand-toggle" href="#">Expand all</a><a class="back-to-top" href="#">Back to top</a><a class="go-to-bottom" href="#">Go to bottom</a></div>
            </ul>
        </div>
    </div>
    <div id="ui-toc-affix" class="ui-affix-toc ui-toc-dropdown unselectable hidden-print" data-spy="affix" style="top:17px;display:none;" null null>
        <div class="toc"><ul class="nav">
<li class=""><a href="#FAIRbioinfo-for-bioinformaticians-2020" title="FAIRbioinfo for bioinformaticians 2020">FAIRbioinfo for bioinformaticians 2020</a><ul class="nav">
<li class=""><a href="#Introduction" title="Introduction">Introduction</a></li>
<li class=""><a href="#Exercice-setup" title="Exercice setup">Exercice setup</a></li>
<li><a href="#Objective-1-the-rule-concept-1-input" title="Objective 1: the rule concept, 1 input">Objective 1: the rule concept, 1 input</a></li>
<li><a href="#Objective-2-the-rule-concept-2-inputs" title="Objective 2: the rule concept, 2 inputs">Objective 2: the rule concept, 2 inputs</a></li>
<li><a href="#Objective-3-expand-function-n-inputs" title="Objective 3: expand function, n inputs">Objective 3: expand function, n inputs</a></li>
<li><a href="#Objective-4-2-rules-no-link" title="Objective 4: 2 rules, no link">Objective 4: 2 rules, no link</a></li>
<li class=""><a href="#Objective-5-The-target-rule" title="Objective 5: The target rule">Objective 5: The target rule</a></li>
<li><a href="#Objective-6-Running-n-individual-jobs" title="Objective 6: Running n individual jobs">Objective 6: Running n individual jobs</a></li>
<li><a href="#Objective-7-Adding-log-file" title="Objective 7: Adding log file">Objective 7: Adding log file</a></li>
<li><a href="#Setup-second-part" title="Setup second part">Setup second part</a><ul class="nav">
<li><a href="#Data" title="Data">Data</a></li>
<li><a href="#Bash-command-line" title="Bash command line">Bash command line</a></li>
<li><a href="#Script-in-3-main-blocks" title="Script in 3 main blocks">Script in 3 main blocks</a></li>
<li><a href="#getops-block" title="getops block">getops block</a></li>
</ul>
</li>
<li><a href="#Objective-8-Adding-a-configuration-file" title="Objective 8: Adding a configuration file">Objective 8: Adding a configuration file</a></li>
<li><a href="#Objective-9-for-block" title="Objective 9: for block">Objective 9: for block</a></li>
<li><a href="#Objective-10-Quality-control-fastqc" title="Objective 10: Quality control, fastqc">Objective 10: Quality control, fastqc</a></li>
<li><a href="#Objective-11-Reads-mapping-bowtie2" title="Objective 11: Reads mapping, bowtie2">Objective 11: Reads mapping, bowtie2</a></li>
<li><a href="#Objective-11-samtools" title="Objective 11: samtools">Objective 11: samtools</a></li>
<li><a href="#Objective-12-FeatureCount" title="Objective 12: FeatureCount">Objective 12: FeatureCount</a></li>
<li><a href="#Objective-13-Count-matrix-creation" title="Objective 13: Count matrix creation">Objective 13: Count matrix creation</a></li>
<li><a href="#DEseq2" title="DEseq2">DEseq2</a></li>
</ul>
</li>
</ul>
</div><div class="toc-menu"><a class="expand-toggle" href="#">Expand all</a><a class="back-to-top" href="#">Back to top</a><a class="go-to-bottom" href="#">Go to bottom</a></div>
    </div>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.1.1/jquery.min.js" integrity="sha256-hVVnYaiADRTO2PzUGmuLJr8BLUSjGIZsDYGmIJLv2b8=" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous" defer></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/gist-embed/2.6.0/gist-embed.min.js" integrity="sha256-KyF2D6xPIJUW5sUDSs93vWyZm+1RzIpKCexxElmxl8g=" crossorigin="anonymous" defer></script>
    <script>
        var markdown = $(".markdown-body");
        //smooth all hash trigger scrolling
        function smoothHashScroll() {
            var hashElements = $("a[href^='#']").toArray();
            for (var i = 0; i < hashElements.length; i++) {
                var element = hashElements[i];
                var $element = $(element);
                var hash = element.hash;
                if (hash) {
                    $element.on('click', function (e) {
                        // store hash
                        var hash = this.hash;
                        if ($(hash).length <= 0) return;
                        // prevent default anchor click behavior
                        e.preventDefault();
                        // animate
                        $('body, html').stop(true, true).animate({
                            scrollTop: $(hash).offset().top
                        }, 100, "linear", function () {
                            // when done, add hash to url
                            // (default click behaviour)
                            window.location.hash = hash;
                        });
                    });
                }
            }
        }

        smoothHashScroll();
        var toc = $('.ui-toc');
        var tocAffix = $('.ui-affix-toc');
        var tocDropdown = $('.ui-toc-dropdown');
        //toc
        tocDropdown.click(function (e) {
            e.stopPropagation();
        });

        var enoughForAffixToc = true;

        function generateScrollspy() {
            $(document.body).scrollspy({
                target: ''
            });
            $(document.body).scrollspy('refresh');
            if (enoughForAffixToc) {
                toc.hide();
                tocAffix.show();
            } else {
                tocAffix.hide();
                toc.show();
            }
            $(document.body).scroll();
        }

        function windowResize() {
            //toc right
            var paddingRight = parseFloat(markdown.css('padding-right'));
            var right = ($(window).width() - (markdown.offset().left + markdown.outerWidth() - paddingRight));
            toc.css('right', right + 'px');
            //affix toc left
            var newbool;
            var rightMargin = (markdown.parent().outerWidth() - markdown.outerWidth()) / 2;
            //for ipad or wider device
            if (rightMargin >= 133) {
                newbool = true;
                var affixLeftMargin = (tocAffix.outerWidth() - tocAffix.width()) / 2;
                var left = markdown.offset().left + markdown.outerWidth() - affixLeftMargin;
                tocAffix.css('left', left + 'px');
            } else {
                newbool = false;
            }
            if (newbool != enoughForAffixToc) {
                enoughForAffixToc = newbool;
                generateScrollspy();
            }
        }
        $(window).resize(function () {
            windowResize();
        });
        $(document).ready(function () {
            windowResize();
            generateScrollspy();
        });

        //remove hash
        function removeHash() {
            window.location.hash = '';
        }

        var backtotop = $('.back-to-top');
        var gotobottom = $('.go-to-bottom');

        backtotop.click(function (e) {
            e.preventDefault();
            e.stopPropagation();
            if (scrollToTop)
                scrollToTop();
            removeHash();
        });
        gotobottom.click(function (e) {
            e.preventDefault();
            e.stopPropagation();
            if (scrollToBottom)
                scrollToBottom();
            removeHash();
        });

        var toggle = $('.expand-toggle');
        var tocExpand = false;

        checkExpandToggle();
        toggle.click(function (e) {
            e.preventDefault();
            e.stopPropagation();
            tocExpand = !tocExpand;
            checkExpandToggle();
        })

        function checkExpandToggle () {
            var toc = $('.ui-toc-dropdown .toc');
            var toggle = $('.expand-toggle');
            if (!tocExpand) {
                toc.removeClass('expand');
                toggle.text('Expand all');
            } else {
                toc.addClass('expand');
                toggle.text('Collapse all');
            }
        }

        function scrollToTop() {
            $('body, html').stop(true, true).animate({
                scrollTop: 0
            }, 100, "linear");
        }

        function scrollToBottom() {
            $('body, html').stop(true, true).animate({
                scrollTop: $(document.body)[0].scrollHeight
            }, 100, "linear");
        }
    </script>
</body>

</html>
